### PR TITLE
Add support for libraries

### PR DESCRIPTION
**Pull request by StanislasBertrand**
This PR : 
* adds logic to synchronize libraries copied on logic from plugins
* factorizes code between MonitoredPlugin and MonitoredLibrary by making them extend a MonitoredFileSystem class

### CODE REVIEW

**Comment by StanislasBertrand** [comment]
https://app.shortcut.com/dataiku/story/94860/add-project-libs-edition-to-dss-pycharm-extension

File: src/main/java/com/dataiku/dss/intellij/MonitoredFilesIndex.java [hunk 1]
```diff
@@ -84,16 +83,14 @@ public synchronized void removeFromIndex(MonitoredRecipeFile monitoredFile) {
...
-        Preconditions.checkNotNull(monitoredLib, "monitoredLib");
-        log.info(String.format("Stop tracking library directory '%s' corresponding to project on instance '%s'.", monitoredLib.baseDir, monitoredLib.library.instance));
-        monitoredLibraries.remove(monitoredLib.baseDir.getCanonicalPath());
+    public synchronized void removeFromIndex(MonitoredFileSystem monitoredFS) {
+        Preconditions.checkNotNull(monitoredFS, "monitoredFS");
+        log.info(String.format("Stop tracking directory '%s' on instance '%s'.", monitoredFS.baseDir, monitoredFS.fsMetadata.instance));
+        if (monitoredFS instanceof MonitoredPlugin) {
+            monitoredPlugins.remove(monitoredFS.baseDir.getCanonicalPath());
+        } else {
+            monitoredLibraries.remove(monitoredFS.baseDir.getCanonicalPath());
```
[the remark this answers is not in the document]
Comment (StanislasBertrand) [reply src/main/java/com/dataiku/dss/intellij/MonitoredFilesIndex.java:93]: fixed [here](https://github.com/dataiku/dss-integration-pycharm/pull/2/commits/557186639029d5adc142c514948046f356a95510)

File: src/main/java/com/dataiku/dss/intellij/SynchronizeWorker.java [hunk 2]
```diff
@@ -311,7 +311,13 @@ private void synchronizeFolder(DSSClient dssClient, MonitoredFileSystem monitore
                 if (monitoredFS instanceof MonitoredPlugin) {
                     fileContent = file.size == 0 ? new byte[0] : dssClient.downloadPluginFile(pluginId, file.path);
                 } else {
-                    fileContent = file.size == 0 ? new byte[0] : dssClient.downloadLibraryFile(pluginId, file.path);
+                    String fileContentString = file.size == 0 ? "" : dssClient.downloadLibraryFile(pluginId, file.path).data;
+                    // Converting back to bytes to factorize code with plugins
+                    if (fileContentString==null || "".equals(fileContentString)) {
```
[the remark this answers is not in the document]
Comment (StanislasBertrand) [reply src/main/java/com/dataiku/dss/intellij/SynchronizeWorker.java:319]: fixed [here](https://github.com/dataiku/dss-integration-pycharm/pull/2/commits/557186639029d5adc142c514948046f356a95510)

File: src/main/java/com/dataiku/dss/intellij/actions/checkout/CheckoutWorker.java [hunk 3]
```diff
@@ -216,11 +216,16 @@ private void checkoutFolder(DssFileSystemMetadata metadata, String id, List<Virt
                 if (metadata instanceof DssPluginMetadata) {
                     fileContent = remoteFile.size == 0 ? new byte[0] : model.serverClient.downloadPluginFile(id, remoteFile.path);
                 } else {
-                    fileContent = remoteFile.size == 0 ? new byte[0] : model.serverClient.downloadLibraryFile(id, remoteFile.path);
+                    String fileContentString = remoteFile.size == 0 ? "" : model.serverClient.downloadLibraryFile(id, remoteFile.path).data;
+                    // Converting back to bytes to factorize code with plugins
+                    if (fileContentString==null || "".equals(fileContentString)) {
```
[the remark this answers is not in the document]
Comment (StanislasBertrand) [reply src/main/java/com/dataiku/dss/intellij/actions/checkout/CheckoutWorker.java]: fixed [here](https://github.com/dataiku/dss-integration-pycharm/pull/2/commits/557186639029d5adc142c514948046f356a95510)

**Comment by StanislasBertrand** [comment]
> Nice job 😄
> 
> Big 👍 for refactoring. Just a remark: If we add support in the future for webapps and wikis, the code will contain a lot of (if x instanceOf y). If we want to avoid that I think we should introduce some more interfaces around MoniteredFileSystem and delegate the dssClient calls to specific implementations for each type of component. We could use factories to get the right implementation. The code in SynchronizedWorker and BackgroundSynchronizer will be much more clearer and free of boiler plate code.
> 
> Other than above, code seems good and it is working as expected during tests.

Indeed, I didnt think of using factories but that seems way more adapted and would indeed simplify the code quite a bit. I think its probably not worth changing now but as you said we should keep it in mind for next iteration on the extension

**Review by ehysi — APPROVED** [review]

File: src/main/java/com/dataiku/dss/intellij/actions/synchronize/SynchronizeDialog.form [hunk 4]
```diff
@@ -13,7 +13,7 @@
           <grid row="0" column="0" row-span="1" col-span="1" vsize-policy="0" hsize-policy="0" anchor="8" fill="0" indent="0" use-parent-layout="false"/>
         </constraints>
         <properties>
-          <text value="Select the Recipes and Plugins to synchronize with DSS:"/>
+          <text value="Select the Recipes Plugins and Libraries to synchronize with DSS:"/>
```
Comment (apichery) [inline src/main/java/com/dataiku/dss/intellij/actions/synchronize/SynchronizeDialog.form:16]: Missing comma?

File: src/main/java/com/dataiku/dss/model/metadata/DssRecipeMetadata.java [hunk 5]
```diff
@@ -1,6 +1,6 @@
 package com.dataiku.dss.model.metadata;
 
-public class DssRecipeMetadata {
+public class DssRecipeMetadata{
```
Comment (apichery) [inline src/main/java/com/dataiku/dss/model/metadata/DssRecipeMetadata.java:3]: Nitpick: I like the whitespace before `{` :)

File: src/main/java/com/dataiku/dss/intellij/actions/synchronize/SynchronizeDialog.form (hunk 4 above)
Comment (StanislasBertrand) [reply src/main/java/com/dataiku/dss/intellij/actions/synchronize/SynchronizeDialog.form:16]: Added in the new PR

File: src/main/java/com/dataiku/dss/model/metadata/DssRecipeMetadata.java (hunk 5 above)
Comment (StanislasBertrand) [reply src/main/java/com/dataiku/dss/model/metadata/DssRecipeMetadata.java:3]: Didnt see this get through, I agree with you, I think modifying existing code just for small style points makes a PR more confusing for minimal gains. Fixed in the new PR

**Pull request merged**